src/data/buffer/parser/basic.lean

```diff
@@ -284,202 +290,203 @@ lemma guard_eq_fail {p : Prop} [decidable p] :
-by simpa [h] using (hp cb n).left
+namespace mono
 
-lemma pure : valid (pure a) :=
-λ _ _, by simp [pure_eq_done]
+instance pure : mono (pure a) :=
+⟨λ _ _, by simp [pure_eq_done]⟩
 
-@[simp] lemma bind {f : α → parser β} (hp : p.valid) (hf : ∀ a, (f a).valid) :
-  (p >>= f).valid :=
+@[simp] instance bind {f : α → parser β} [p.mono] [∀ a, (f a).mono] :
+  (p >>= f).mono :=
 begin
+  constructor,
   intros cb n,
   cases hx : (p >>= f) cb n,
   { obtain ⟨n', a, h, h'⟩ := bind_eq_done.mp hx,
-    refine ⟨le_trans (hp cb n).left (h' ▸ _), λ hn, (hp cb n).right _⟩,
-    { simp [h, hf a cb n'] },
-    { simp [h, (hf a cb n').right (h'.symm ▸ hn)] } },
+    refine le_trans (of_done h) _,
+    simpa [h'] using of_done h' },
   { obtain h | ⟨n', a, h, h'⟩ := bind_eq_fail.mp hx,
-    { simpa only [h] using hp cb n },
-    { refine ⟨le_trans (hp cb n).left (h' ▸ _), λ hn, (hp cb n).right _⟩,
-      { simp [h, hf a cb n'] },
-      { simp [h, (hf a cb n').right (h'.symm ▸ hn)] } } }
+    { simpa [h] using of_fail h },
+    { refine le_trans (of_done h) _,
+      simpa [h'] using of_fail h' } }
 end
 
-lemma and_then {q : parser β} (hp : p.valid) (hq : q.valid) : (p >> q).valid :=
-hp.bind (λ _, hq)
+instance and_then {q : parser β} [p.mono] [q.mono] : (p >> q).mono := mono.bind
 
-@[simp] lemma map (hp : p.valid) {f : α → β} : (f <$> p).valid :=
-hp.bind (λ _, pure)
+instance map [p.mono] {f : α → β} : (f <$> p).mono := mono.bind
 
-@[simp] lemma seq {f : parser (α → β)} (hf : f.valid) (hp : p.valid) : (f <*> p).valid :=
-hf.bind (λ _, hp.map)
+instance seq {f : parser (α → β)} [f.mono] [p.mono] : (f <*> p).mono := mono.bind
 
-@[simp] lemma mmap {l : list α} {f : α → parser β} (h : ∀ a ∈ l, (f a).valid) :
-  (l.mmap f).valid :=
-begin
-  induction l with hd tl hl generalizing h,
-  { exact pure },
-  { exact bind (h _ (list.mem_cons_self _ _))
-      (λ b, map (hl (λ _ ha, h _ (list.mem_cons_of_mem _ ha)))) }
+instance mmap : Π {l : list α} {f : α → parser β} [∀ a ∈ l, (f a).mono],
+  (l.mmap f).mono
+| []       _ _ := mono.pure
+| (a :: l) f h := begin
+  convert mono.bind,
+  { exact h _ (list.mem_cons_self _ _) },
+  { intro,
+    convert mono.map,
+    convert mmap,
+    exact (λ _ ha, h _ (list.mem_cons_of_mem _ ha)) }
 end
 
-@[simp] lemma mmap' {l : list α} {f : α → parser β} (h : ∀ a ∈ l, (f a).valid) :
-  (l.mmap' f).valid :=
-begin
-  induction l with hd tl hl generalizing h,
-  { exact pure },
-  { exact and_then (h _ (list.mem_cons_self _ _)) (hl (λ _ ha, h _ (list.mem_cons_of_mem _ ha))) }
+instance mmap' : Π {l : list α} {f : α → parser β} [∀ a ∈ l, (f a).mono],
+  (l.mmap' f).mono
+| []       _ _ := mono.pure
+| (a :: l) f h := begin
+  convert mono.and_then,
+  { exact h _ (list.mem_cons_self _ _) },
+  { convert mmap',
+    exact (λ _ ha, h _ (list.mem_cons_of_mem _ ha)) }
 end
 
-@[simp] lemma failure : @parser.valid α failure :=
-by simp [valid, le_refl]
+instance failure : @parser.mono α failure :=
```

Comment:
I find this style to be easier to read, but it's equivalent and just personal preference, so up to you if you want to change:
```suggestion
instance failure : (failure : parser α).mono :=
```